snakemake/uf_cells_spleen_0510/src/generate_cell_sets.py

```diff
@@ -0,0 +1,170 @@
+        assert(CL_ROOT_ID in ancestor_term_set)
+
+        # Initialize the current node ID to the cell type of interest
+        curr_node_id = node_id
+
+        # Construct a list of ancestors, inclusive.
+        # [node_id, parent_id, grandparent_id, ...]
+        ancestors = [curr_node_id]
+
+        # Get the parents of the current node.
+        curr_node_parents = list(graph.out_edges(curr_node_id, keys=True))
+
+        # Warn if the current node has multiple parents.
+        if len(curr_node_parents) > 1:
+            print((
+                f"WARN: {curr_node_id} has "
+                f"{len(curr_node_parents)} parents."
+            ))
+
+        # Select the first (hopefully only) parent node.
+        _, curr_parent_id, relationship = curr_node_parents[0]
+        assert(relationship == "is_a")
+
+        while curr_node_id != CL_ROOT_ID:
+            # Get the parents of the current node.
+            curr_node_parents = list(graph.out_edges(curr_node_id, keys=True))
+
+            # Warn if the current node has multiple parents.
+            if len(curr_node_parents) > 1:
+                print((
+                    f"WARN: {curr_node_id} has "
+                    f"{len(curr_node_parents)} parents."
+                ))
+
+            # Select the first (hopefully only) parent node.
+            _, curr_parent_id, relationship = curr_node_parents[0]
+            assert(relationship == "is_a")
+            ancestors.append(curr_parent_id)
+
+            # Set the current node to its parent to
+            # prepare for the next iteration.
+            curr_node_id = curr_parent_id
+
+        named_ancestors = [id_to_name[a] for a in ancestors]
+        named_ancestors_reversed = list(reversed(named_ancestors))
+        print(named_ancestors_reversed)
+
+        ancestors_and_sets.append((
+            named_ancestors_reversed,
+            cell_type_df[COLUMNS.CELL_ID.value].unique().tolist()
+        ))
+
+    # Pop off all ancestors that are the same for all cell types.
+    # e.g. 'cell', 'native cell', ...
+    ancestor_list_lens = [len(x[0]) for x in ancestors_and_sets]
+    min_ancestor_list_len = min(ancestor_list_lens)
+    assert(min_ancestor_list_len >= 1)
+    for level in range(min_ancestor_list_len - 1):
+        unique_level_cell_types = set()
+        for ancestors, cell_set in ancestors_and_sets:
+            unique_level_cell_types.add(ancestors[0])
+
+        if len(unique_level_cell_types) == 1:
+            for ancestors, cell_set in ancestors_and_sets:
+                ancestors.pop(0)
+        else:
+            print(unique_level_cell_types)
+            break
+
+    # Construct a hierarchy of cell types.
+    def find_or_create_parent(d, keys, child):
+        key = keys[0]
+
+        if key in d and isinstance(d[key], dict):
+            result = d[key]
+        else:
+            result = d[key] = dict()
+
+        if len(keys) == 1:
+            result["any"] = child
```

Comment:
Why can't we just put this in the parent?  It looks like you could do that from the outputs.

Reply:
This is something that probably requires a larger discussion. Right now there are  annotations that associate with non-leaf nodes.

For example, in the dataset here there are some cells annotated as just "B cell" and other cells annotated as "B cell, CD19-positive" whose parent is "B cell". In the schema I developed, only leaf nodes can have 'sets' associated, which is why I added the extra "any" node (i.e. to ensure that all sets are in a leaf node).

Reply:
- B cell
    - `set` [cell_1, cell_2, cell_3]
    - `children`
        - B cell, CD19-positive
            - `set` [cell_4, cell_5, cell_6]

becomes

- B cell
    - `children`
        - any
            - `set` [cell_1, cell_2, cell_3]
        - B cell, CD19-positive
            - `set` [cell_4, cell_5, cell_6]

Reply:
Ok.  That makes sense.